### ISSUE TITLE
Figured out why Bert in Shark tank performs much better than Bert from torchbench

### ISSUE THREAD

**Issue by alliemartens** [iree-org/iree#10110]

**Comment by alliemartens** [comment]
Adding Ramiro's notes here.

Input and vocab size are different 
- Shark: input_size=128, vocab_size=2
- Torchbench: input_size=512, vocab_size=30522

If I change torchbench parameters to match Shark, I get IREE 2x faster than PyTorch eager

Using input_size=512 and vocab_size that is of same order as 30522 but 32 aligned makes IREE perform 2x slower than PyTorch eager

**Comment by alliemartens** [comment]
@ThomasRaoux @mattwalsh @rsuderman @KoolJBlack @mariecwhite @silvasean for vis

**Comment by mariecwhite** [comment]
A few things to double-check to make sure the models are close to apple-to-apples:
- Since the Shark model has smaller input and vocab size, it would suggest that the model is smaller. For example, BERT-base and BERT-large vary in the number of layers, number of attention heads and number of hidden layers (12 layers vs 24 layers, 12 attention heads vs 16 attention heads, 768 hidden layers vs 1024 hidden layers). So even if the input size and vocab size are changed to match, the internals may still differ greatly.
- Does changing the input size and vocab size change the internals as expected? Some rules of thumb: i) BERT scales quadratically according to sequence length (input_size), ii) As the vocab size increases, the matrices in the input and output layers should also increase.
- BERT latency depends on input and thus how it's trained. Are the models pre-trained (on a generic dataset) or fine-tuned (for a specific task e.g. question-answering). We would need to make sure that both models have been trained similarly and the input data is the same.

**Comment by ramiro050** [comment]
A correction on the update I gave about the differences between `bert-base-uncased` in Shark vs. the one used by torchbench. Both models do actually use the same vocabulary size, 30522 (Shark does use a smaller input size: 128 vs 512). The difference is that the Shark model is doing sequence classification outputting classificatioon scores on two labels https://huggingface.co/docs/transformers/v4.21.1/en/model_doc/bert#transformers.BertForSequenceClassification, while the torchbench model is doing a masked language modeling outputting prediction scores for each of the 30522 vocabulary tokens https://huggingface.co/docs/transformers/v4.21.1/en/model_doc/bert#transformers.BertForMaskedLM.

The difference between the two is only at the end of the model. Because the Shark model is only doing classification into two labels, the final op has a matmul between tensors of size `1x768`, `768x2`. In other words, the model never does a computation with a tensor that has a dimension of size 30522. On the other hand, the torchbench model performs a matmul at the end between tensors of size `512x768`, `768x30522`, which is the main source of the performance difference between the two. So the two models are actually achieving different goals.

**Comment by ramiro050** [comment]
All configuration parameters used between the two models are the same:

Shark:

```
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "gradient_checkpointing": false,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "torchscript": true,
  "transformers_version": "4.12.1",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
```

Torchbench:

```
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "initializer_range": 0.02,
  "intermediate_size": 3072,
  "layer_norm_eps": 1e-12,
  "max_position_embeddings": 512,
  "model_type": "bert",
  "num_attention_heads": 12,
  "num_hidden_layers": 12,
  "pad_token_id": 0,
  "position_embedding_type": "absolute",
  "transformers_version": "4.12.1",
  "type_vocab_size": 2,
  "use_cache": true,
  "vocab_size": 30522
```

**Comment by mariecwhite** [comment]
@powderluv, @monorimet  Given the differences, can we update the Shark model to be the same as what is in Torch bench?

**Comment by monorimet** [comment]
Sounds good. Ill take care of this.

**Comment by erob710** [comment]
Confirm model in Torch is same as shark tank

**Comment by benvanik** [comment]
Is this issue still valid? Sounds like they're the same now?

**Comment by silvasean** [comment]
> Confirm model in Torch is same as shark tank

Was this a statement or a request?

**Comment by mariecwhite** [comment]
That was a request. I've looked into this and SHARK and TorchDynamo now look the same except for the sequence length. See https://github.com/nod-ai/SHARK/issues/324 for details. I'll close this and we can track it in the other issue.
